[#12557] Instructor edit feedback session page: NumberFormatException when inputting decimal numbers into distribute points questions

dlimyy opened the pull request:
Fixes #12557

**Outline of Solution**

Similar to the numerical scale question, I made it such that the number inputted will always be an integer so if the user tries keying in a number such as 99.5, it will be rounded up to 100. This helps to prevent this error from occurring and also reduces amount of code which has to be changed to accommodate decimals.

Screen recording of solution
https://github.com/TEAMMATES/teammates/assets/97420966/a311869e-e8e0-40a7-8724-cc568cb9e27d

weiquu reviewed (commented):
Hi @dlimyy, fix looks good! I would suggest shifting the `ceil` implementation into the `QuestionEditDetailsFormComponent` base class (`question-edit-details-form.component.ts`), since it now can be used for 3 different question types. Then, you can delete the 2 functions in the constsum classes, as well as edit the numscale html file to use the `ceil` function.

dlimyy commented:
Thank you for reviewing my PR and sorry for the delay in making these changes. I have shifted ceils' implementation to the base class and edited numscale's html file.

dlimyy commented:
Ready for review

dlimyy commented:
Thanks for reviewing my PR and sorry for the delay in response. I agree with you that we should be more consistent in the behaviour of the input validation so I have adopted the (keypress) solution mentioned above. However, I discovered that this solution does not prevent the user from pasting decimals numbers in. To resolve this, I implemented the onPaste method which restricts the content that can be pasted into the fields to be strictly numbers only.

List of changes made from previous PR:
* Added the onPointsInput as suggested above
* Added the onPaste method to prevent non-numerical inputs from being pasted
* Added max to prevent user from clicking the number spinner wheel to a number that will cause the error in this PR

dlimyy commented on src/web/app/components/question-types/question-edit-details-form/constsum-options-question-edit-details-form.component.html:
```diff
@@ -34,8 +34,8 @@
             <div class="col-sm-6">
               <input id="total-points-radio" class="form-check-input" type="radio" [ngModel]="model.pointsPerOption" [value]="false"
                      (ngModelChange)="triggerModelChange('pointsPerOption', $event)" [disabled]="!isEditable" aria-label="Use Total Points Checkbox">
-              <input id="total-points" type="number" class="form-control" min="1" step="1"
```
I removed it initially as I noticed that it does not restrict the input that the user keys in. However, I just discovered  that it is used for the number spinner input range so I have added it back

cedricongjh commented on src/web/app/components/question-types/question-edit-details-form/question-edit-details-form.component.ts:
```diff
@@ -38,4 +38,26 @@ export abstract class QuestionEditDetailsFormComponent<D extends FeedbackQuestio
     this.detailsChange.emit({ ...this.model, ...obj });
   }
+
+  onPointsInput(event: KeyboardEvent): void {
+    const { key } = event;
+    const isBackspace = key === 'Backspace';
+    const isDigit = /[0-9]/.test(key);
+    if (!isBackspace && !isDigit) {
+      event.preventDefault();
+    }
+  }
+
+  onPaste(event: ClipboardEvent): void {
+    const { clipboardData } = event;
+    if (clipboardData == null) {
+      return;
+    }
+    const pastedText = clipboardData.getData('text');
+    const isDigit = /^\d+$/.test(pastedText);
+    if (!isDigit) {
+      event.preventDefault();
+    }
```
hey @dlimyy, I think it would be good if we add tests for these new functions

dlimyy replied:
Hi @cedricongjh, sorry for getting back to you late! I will add the tests for these new functions soon.

dlimyy commented:
I have added the tests for onPointsInput but currently facing some difficulties in adding the test for the onPasteMethod. When trying to write the test for the method(as shown below), I constantly face the error `ReferenceError: ClipboardEvent is not defined`. I am not really able to find a solution to this and this error does not occur for other events such as `KeyboardEvent` and `MouseEvent`. 
```
 it('should prevent alphabetical input in onPaste', () => {
    const event : ClipboardEvent = new ClipboardEvent('paste');

    event.clipboardData?.setData('text/plain', 'abc');
    const preventDefaultSpy = spyOn(event, 'preventDefault');
    component.onPaste(event);
    expect(preventDefaultSpy).not.toHaveBeenCalled();
  });

```

I am wondering if anyone knows how to resolve this error. Thanks!

cedricongjh commented:
hi @dlimyy, I've looked into the issue, and I found that  `ClipboardEvent` is unfortunately not accessible by the test environment. 

A solution I've found to mock a paste event is as follows:


```
    const inputElement = fixture.debugElement.query(By.css('#max-value')).nativeElement as HTMLInputElement;
    inputElement.value = 'abc';
    expect(inputElement.value).toBe('')

    inputElement.value = '2333';
    expect(inputElement.value).toBe('2333')

```

Do try it out and let me know if it works!

dlimyy commented:
Hi @cedricongjh, thanks for your help and suggestion! I have tested this out on my end but it does not seem to invoke the onPaste method in the process. If I changed `inputElement.value = "2.2"`, it returns `2.2` instead of `''`. Since the node.js testing environment does not provide support for this ClipboardEvent, would it ok in this case to leave this method untested?

cedricongjh commented:
> Hi @cedricongjh, thanks for your help and suggestion! I have tested this out on my end but it does not seem to invoke the onPaste method in the process. If I changed `inputElement.value = "2.2"`, it returns `2.2` instead of `''`. I believed the reason why it worked for the 'abc' case is that the input type of `max-value` is number so any alphabetical characters passed in would get cancelled out. Since the node.js testing environment does not provide support for this ClipboardEvent, would it ok in this case to leave this method untested?

Ah that's unfortunate, let's leave that untested for now

cedricongjh reviewed (changes requested):
Hey @dlimyy, seems like i'm able to enter very large numbers: 
![image](https://github.com/TEAMMATES/teammates/assets/67156011/6bdb58e5-91fd-4ad0-8988-3c9ccd4c140a)

Do let me know if you can replicate this on your end and fix it!

dlimyy commented:
Hi @cedricongjh, sorry for the delay in getting back to you! I am also able to replicate this bug on my end so currently I am working on a solution which restricts the length of the number to 9 characters to eliminate the chance of the bug occurring

dlimyy commented:
I have added my solution which restricts the user from keying in a number that has a length greater than 9 characters. The assignment of `target.value` is needed to ensure that the number displayed will be correct though the actual points in the model is correct. Could I check if my solution looks ok? I will work on adding the tests for these 3 functions.

cedricongjh commented:
> I have added my solution which restricts the user from keying in a number that has a length greater than 9 characters. The assignment of `target.value` is needed to ensure that the number displayed will be correct though the actual points in the model is correct. Could I check if my solution looks ok? I will work on adding the tests for these 3 functions.

hey @dlimyy I think it'll be much better if we added the checking for numbers that are too long in the `onPointsInput` method, we could do this by accessing the value of the input element like this:

```
(event.target as HTMLInputElement).value
```

dlimyy commented:
Hi @cedricongjh, thank you for your suggestion. I have initially considered and tried implementing this in onPointsInput but I noticed that this would have a few issues. Firstly, if the user highlights a 9-digit number and types a number over it, the expected behaviour is that the highlighted/selected region is replaced with a number. However, if implemented in onPointsInput, only the last number would change rather than the whole highlighted region. Another issue would be that this does not prevent the user from pasting a larger input(Implementing this in the onPaste region with the highlighted region is also quite complicated).

As I noted that my previous solution was quite clunky(with the methods implemented in the individual components), I decided to instead abstract it out into the `question-edit-details-form.component`(similar to what I did for onPointsInput and onPaste) and found a way for it to update the model and display correctly while ensuring that the aforementioned highlighting problem does not occur. 

I have also added the tests for the `restrictPointsLength` method as well.

cedricongjh reviewed (approved):
LGTM, thank you for the work on this!

jasonqiu212 reviewed (commented):
Hi @dlimyy, Sorry for the late review! The fix looks good for the 2 distribute points questions. I do have some reservations for the numeric scale question. Since we still want a max cap on the inputs, I'm still not too sure what's a good solution

dlimyy commented:
Hi @jasonqiu212, thanks for reviewing my PR. For the numerical scale question case, would changing the stepper range to `min = 1` and `max = 999999999 ` be a solution since it ensures that the range of the stepper is within value and if the user selects something less than minimum value in the maximum value box, there is an existing error message to catch this.

jasonqiu212 commented:
Hi @dlimyy, Ah yes, I think that's a valid solution. We can indeed just make use of the existing error message

dlimyy commented:
Hi @jasonqiu212, I have implemented this solution to fix this bug.

jasonqiu212 commented on src/web/app/components/question-types/question-edit-details-form/question-edit-details-form.component.ts:
```diff
@@ -38,4 +38,33 @@ export abstract class QuestionEditDetailsFormComponent<D extends FeedbackQuestio
     this.detailsChange.emit({ ...this.model, ...obj });
   }
+
+  onPointsInput(event: KeyboardEvent): void {
```
Since we're using this method for the max and min value inputs in numeric scale, I think this method requires a more general name. Perhaps `onIntegerInput`?

jasonqiu212 commented on src/web/app/components/question-types/question-edit-details-form/question-edit-details-form.component.ts:
```diff
@@ -38,4 +38,33 @@ export abstract class QuestionEditDetailsFormComponent<D extends FeedbackQuestio
+    }
+  }
+
+  restrictPointsLength(event : InputEvent, field: keyof D) : void {
```
Similar reason to above, I think this method requires a more general name. Perhaps `restrictIntegerInputLength`?

jasonqiu212 commented on src/web/app/components/question-types/question-edit-details-form/num-scale-question-edit-details-form.component.html:
```diff
@@ -10,7 +11,8 @@
   </div>
```
Currently, the increment step input has no maximum cap. Similar to the other inputs, a big number may also cause problems in the backend. I understand that this input has 1 difference with the rest - which is it allows decimal input. However, I think it's a bit odd for this input to allow user inputs with more than 9 digits.

What we can do is:
- Add a `max="999999999"` to increment step input
- Create a new method called `restrictFloatInputLength` to restrict **decimal** inputs to 9 digits (inc. the decimal point). The current `restrictPointsLength` method does not support decimal inputs, since it performs `parseInt`. See below comment on my suggestions for implementing this new method
- Add a `(input)="restrictFloatInputLength($event, 'step')"` to increment step input

jasonqiu212 commented on src/web/app/components/question-types/question-edit-details-form/question-edit-details-form.component.ts:
```diff
@@ -38,4 +38,33 @@ export abstract class QuestionEditDetailsFormComponent<D extends FeedbackQuestio
+      target.value = target.value.substring(0, 9);
+      this.triggerModelChange(field, parseInt(target.value, 10) as any);
+    }
+  }
```
`restrictFloatInputLength` can follow the exact same implementation as `restrictPointsLength` method, except: `parseInt(target.value, 10)` can be changed to `parseFloat(target.value)`

jasonqiu212 commented on src/web/app/components/question-types/question-edit-details-form/num-scale-question-edit-details-form.component.spec.ts:
```diff
@@ -25,4 +26,53 @@ describe('NumScaleQuestionEditDetailsFormComponent', () => {
+    component.restrictPointsLength(inputEvent, 'minScale');
+    expect((inputEvent.target as HTMLInputElement).value).toEqual('123456789');
+  });
+
```
If we decide to go with the below suggestions, let's add a test for the new method too

jasonqiu212 reviewed (changes requested):
Hi @dlimyy, I apologize for the excessive back and forth on this PR! Really a big props to you for taking on this tough issue!

I mainly have 1 more suggestion on the increment step input for the numerical scale question. Please let me know what you think or any considerations you may have too!

dlimyy commented:
Hi @jasonqiu212, I have added the changes mentioned above. I have also added a new method called `onFloatInput` as it is needed to prevent the user from keying in the scientific notation `e`.

jasonqiu212 commented:
Just a note for future reference, custom methods `restrictIntegerInputLength` and `restrictFloatInputLength` was necessary for the following reasons:
- `maxlength` attribute for `<input>` does not work `type="number"`
- `min` and `max` attributes for `<input>` were not sufficient, as the restrictions only apply when clicking on the stepper (i.e. User can still input numbers greater than `max`)
